File: src/Python/227.BasicCalculatorII.py

class Solution:
    def calculate(self, s):
        s += "+0"
        stack = []
        num = 0
        sign = "+"
        for c in s:
            if c.isdigit():
                num = num * 10 + int(c)
            elif c in {"+", "-", "*", "/"}:
                if sign == "+":
                    stack.append(num)
                elif sign == "-":
                    stack.append(-num)
                elif sign == "*":
                    stack[-1] = stack[-1] * num
                elif sign == "/":
                    if stack[-1] < 0:
                        stack[-1] = -(-stack[-1] // num)
                    else:
                        stack[-1] = stack[-1] // num
                sign, num = c, 0
        return sum(stack)


if __name__ == "__main__":
    solution = Solution()
    print(solution.calculate("3+2*2"))
    print(solution.calculate(" 3/2 "))
    

File: src/CSharp/227.BasicCalculatorII.cs

using System;
using System.Collections.Generic;

namespace PeefyLeetCode.BasicCalculatorII
{
public class Solution {
    public int Calculate(string s) {
        Stack<int> st = new Stack<int>();
        int n = s.Length, num = 0, res = 0;
        char op = '+';
        for (int i = 0; i < n; ++i){
            if (s[i] >= '0')
                num = num * 10 + (s[i] - '0');
            if ((s[i] < '0' && s[i] != ' ') || i == n - 1){
                if (op == '+')
                    st.Push(num);
                if (op == '-')
                    st.Push(-num);
                if (op == '*' || op == '/'){
                    int temp = (op == '*') ? st.Peek() * num : st.Peek() / num;
                    st.Pop();
                    st.Push(temp);
                }
                op = s[i];
                num = 0;
            }
        }
        while (st.Count > 0){
            res += st.Pop();
        }
        return res;
    }
}
}